### PR TITLE
Use Array.Empty<T> in ToArray

### LINKED ISSUE #2363: Should ToArray methods be allowed to return Array.Empty<T>()?
We've been trying to use ```Array.Empty<T>``` in most places where we would otherwise construct a new ```T[0]```.  Some places we've explicitly tried to stay away from changing are ```ToArray``` methods, in particular because of an assumption that there's an expection that these methods always produce a new array, e.g. with MSDN documentation that states things like "Creates an array", "Copies the elements to a new array", etc.

The question is: should we relax that and instead use ```Array.Empty<T>``` in ToArray methods as well?

Several months back, before the corefx libraries were on GitHub, I'd done a sweep through them doing hundreds of empty array replacements.  In general I stayed away from changing anything that looked like it was risky, e.g. ToArray methods, places where the arrays were obviously used for reference equality, etc.  I apparently was a little over aggressive with regards to my own guidelines, and ended up changing some of the non-generic collection's ToArray methods to use ```Array.Empty```, though.

Personally I'm tempted to say that ToArray methods should be able to use ```Array.Empty```, and that we should update the generic collections and LINQ to do so, but that since the non-generic collections exist primarily for compatibility reasons, we should revert the changes to those ToArray methods.

Thoughts?

### PR DESCRIPTION
**Pull request by stephentoub**
Changing System.Collections, System.Collections.Concurrent, and System.Linq accordingly.

Fixes #2363 
cc: @ellismg, @mellinoe, @VSadov, @JonHanna, @KrzysztofCwalina

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 1]
```diff
@@ -3931,12 +3959,12 @@ public TElement LastOrDefault()
...
         public TElement[] ToArray()
         {
-            return new TElement[0]; // consider replacing with Array.Empty<TElement>()
+            return Array.Empty<TElement>();
         }
 
         public List<TElement> ToList()
         {
-            return new List<TElement>(0);
+            return new List<TElement>();
```
Comment (mellinoe) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Why this change?

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 2]
```diff
@@ -4671,7 +4702,7 @@ internal Buffer(IEnumerable<TElement> source)
 
         internal TElement[] ToArray()
         {
-            if (count == 0) return new TElement[0]; // consider replacing with Array.Empty<TElement>()
+            if (count == 0) return Array.Empty<TElement>();
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: For just about every path, if `count == 0` then `array` should already be `Array.Empty<TElement>()`, so we can just cut this line and have the next line cover it.

**Comment by mellinoe** [comment]
LGTM

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Because the default is 0, but passing 0 forces it to do a few extra checks to validate the argument, to no additional benefit.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: It's only in there because back when .NET was closed-source I picked up the habit of using `(0)` when creating a list that was expected to remain empty, because it seems like a reasonable choice from the outside, though examining the source shows `()` is slightly more efficient. It's reasonable to tidy it up while optimising the memory use of the similar `ToArray()`.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Makes sense.  I'll remove the line.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 3]
```diff
@@ -4044,14 +4072,17 @@ public IEnumerator<TElement> GetEnumerator()
...
             Buffer<TElement> buffer = new Buffer<TElement>(source);
+
             int count = buffer.count;
-            TElement[] array = new TElement[count];
-            if (count > 0)
+            if (count == 0)
             {
-                int[] map = SortedMap(buffer);
-                for (int i = 0; i != array.Length; i++) array[i] = buffer.items[map[i]];
+                return Array.Empty<TElement>();
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: `return buffer.items`? It'll point to the cached empty in this case, and that pointer should be more likely to be in CPU cache, I think.

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: (If my thinking there is reasonable, there are a few other cases it would apply to).

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 4]
```diff
@@ -3931,12 +3959,12 @@ public TElement LastOrDefault()
 
         public TElement[] ToArray()
         {
-            return new TElement[0]; // consider replacing with Array.Empty<TElement>()
+            return Array.Empty<TElement>();
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: If the decision to use `Array.Empty` in these cases had been made before I created this class, I probably would have used its enumerator for the `GetEnumerator()` of this class and not bothered with its implementing `IEnumerator`. Not sure whether it's worth changing to that now, though.

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please.

**Comment by JonHanna** [comment]
A couple of suggestions that may or may not have merit, but LGTM.

Is this being done on coreclr too?

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs [hunk 5]
```diff
@@ -255,8 +255,12 @@ public bool IsEmpty
         /// cref="ConcurrentQueue{T}"/>.</returns>
         public T[] ToArray()
         {
-            return ToList().ToArray();
+            List<T> list = ToList();
```
Comment (KrzysztofCwalina) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs]: Does this create a list even if empty array will be returned?

Comment (VSadov) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs]: It may make sense to call GetHeadTailPositions directly and check if queue is (was, really) empty and return empty. We can assume that ToArray will be typically called in quiescent state, so GetHeadTailPositions should be fairly cheap.

If this same optimization is applied to List<T>.ToArray(), then we do not really need to check the list for emptiness.

**Comment by VSadov** [comment]
LGTM

File: src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs]: > Does this create a list even if empty array will be returned?

Yes.  The changes necessary to avoid that didn't seem worthwhile.  If we really want to, we can do it separately; it would have an impact on some of the synchronization being done, and it would require more scrutiny.

Comment (stephentoub) [inline src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentQueue.cs]: > we do not really need to check the list for emptiness.

Good point.  I'll revert this.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Let's leave it for now.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 6]
```diff
@@ -1827,7 +1844,13 @@ public override void Dispose()
 
             public int[] ToArray()
             {
-                int[] array = new int[_end - _start];
+                int length = _end - _start;
+                if (length == 0)
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Is this reachable? I'm on a phone so I can't check fully, but I'm pretty sure zero size should cause Empty partition to be returned instead of an instance of this class

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 7]
```diff
@@ -1936,6 +1959,11 @@ public override bool MoveNext()
 
             public TResult[] ToArray()
             {
+                if (_count == 0)
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Likewise, is this reachable?

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 8]
```diff
@@ -4093,7 +4124,7 @@ internal TElement[] ToArray(int minIdx, int maxIdx)
         {
             Buffer<TElement> buffer = new Buffer<TElement>(source);
             int count = buffer.count;
-            if (count <= minIdx) return new TElement[0];
+            if (count <= minIdx) return Array.Empty<TElement>();
```
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: On the stack in buffer.items

**Comment by JonHanna** [comment]
Test Code Coverage

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 8 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: How do we know that items is empty?

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I don't believe so.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Nope.

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Well, actually, it is, but only due to a bug.  Can you spot it? :smile: (https://github.com/dotnet/corefx/pull/5562/files#diff-95a14d3b4ae14339b623dd95c7537fd0R1923)  I'm fixing that, too.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 8 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I "know" because I'm mis-reading! Ignore that comment.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 7 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: That was nasty.

**Comment by stephentoub** [comment]
Thanks, all, for the reviews.
